src/core/ext/client_channel/retry_throttle.c

```diff
@@ -0,0 +1,234 @@
+
+#include <limits.h>
+#include <string.h>
+
+#include <grpc/support/alloc.h>
+#include <grpc/support/atm.h>
+#include <grpc/support/avl.h>
+#include <grpc/support/string_util.h>
+#include <grpc/support/sync.h>
+
+//
+// server_retry_throttle_data
+//
+
+struct grpc_server_retry_throttle_data {
+  gpr_refcount refs;
+  int max_milli_tokens;
+  int milli_token_ratio;
+  gpr_atm milli_tokens;
+  // A pointer to the replacement for this grpc_server_retry_throttle_data
+  // entry.  If non-NULL, then this entry is stale and must not be used.
+  // We hold a reference to the replacement.
+  gpr_atm replacement;
+};
+
+static void get_replacement_throttle_data_if_needed(
+    grpc_server_retry_throttle_data** throttle_data) {
+  while (true) {
+    grpc_server_retry_throttle_data* new_throttle_data =
+        (grpc_server_retry_throttle_data*)gpr_atm_acq_load(
+            &(*throttle_data)->replacement);
+    if (new_throttle_data == NULL) return;
+    *throttle_data = new_throttle_data;
+  }
+}
+
+bool grpc_server_retry_throttle_data_record_failure(
+    grpc_server_retry_throttle_data* throttle_data) {
+  // First, check if we are stale and need to be replaced.
+  get_replacement_throttle_data_if_needed(&throttle_data);
+  // We decrement milli_tokens by 1000 (1 token) for each failure.
+  const int delta = -1000;
+  const int old_value = (int)gpr_atm_full_fetch_add(
+      &throttle_data->milli_tokens, (gpr_atm)delta);
+  // If the above change takes us below 0, then re-add the excess.  Note
+  // that between these two atomic operations, the value will be
+  // artificially low by as much as 1000, but this window should be
+  // brief.
+  int new_value = old_value - 1000;
+  if (new_value < 0) {
+    const int excess_value = new_value - (old_value < 0 ? old_value : 0);
+    gpr_atm_full_fetch_add(&throttle_data->milli_tokens,
+                           (gpr_atm)-excess_value);
+    new_value = 0;
+  }
+  // Retries are allowed as long as the new value is above the threshold
+  // (max_milli_tokens / 2).
+  return new_value > throttle_data->max_milli_tokens / 2;
+}
+
+void grpc_server_retry_throttle_data_record_success(
+    grpc_server_retry_throttle_data* throttle_data) {
+  // First, check if we are stale and need to be replaced.
+  get_replacement_throttle_data_if_needed(&throttle_data);
+  // We increment milli_tokens by milli_token_ratio for each success.
+  const int delta = throttle_data->milli_token_ratio;
+  const int old_value = (int)gpr_atm_full_fetch_add(
+      &throttle_data->milli_tokens, (gpr_atm)delta);
+  // If the above change takes us over max_milli_tokens, then subtract
+  // the excess.  Note that between these two atomic operations, the
+  // value will be artificially high by as much as milli_token_ratio,
+  // but this window should be brief.
+  const int new_value = old_value + throttle_data->milli_token_ratio;
+  if (new_value > throttle_data->max_milli_tokens) {
+    const int excess_value =
```

Comment:
Unless I'm missing something (which is entirely possible), I think the current implementation will actually work fine in that case.

For example, let's say that max_tokens is 10, token_ratio is 0.5, and the current value is 9.75.  Now we have the following steps:

1. First success increments by 0.5: old_value=9.75, new_value=10.25.
2. Second success increments by 0.5: old_value=10.25, new_value=10.75
3. First success computes excess: old_value < 10, so excess = new_value (10.25) - max_tokens (10) == 0.25; subtracting 10.75 - 0.25 == 10.5
4. Second success computes excess: old_value > 10, so excess = new_value (10.75) - old_value (10.25) == 0.5; subtracting 10.5 - 0.5 == 10

I don't think it matters what order steps 3 and 4 occur in, because the net result of the two subtractions will always take us back to the max value, regardless of the order.

Can you give me an example of a case where this won't work?

(I'm certainly happy to change this to use a CAS loop if necessary, but my understanding is that that's less efficient, so it seems better to avoid it unless absolutely necessary.)